### DISCUSSION TITLE
New instance per request or per test worker? (Vitest)

### DISCUSSION THREAD

**Discussion by typedarray** [wagmi-dev/anvil.js#16] [Q&A] [answered]
I'm planning to adopt anvil.js with Vitest, but it's not clear to me how new Anvil instances are spawned when using `startProxy`. Is it truly per request as the docs suggest, or is it per vitest test worker?

Seems like there are several possible ways that this might behave:
1) Fresh instance per (RPC) request
2) Fresh instance per test
3) Fresh instance per test worker

I also imagine this depends on the Vitest [threads](https://vitest.dev/config/#threads) config. Appreciate any additional info - thanks for building this!

**Comment by fubhy** [comment, marked as answer]
It's per `id`. And since you'd use one `id` per test thread, it's per test essentially.

With this, you can completely drop the `threads` option (default is `true`) and simply let `vitest` decide how many cores to utilize for your test runs in parallel.

Check out the example here:

This sets up globalSetup.ts (see vitest.config.ts). It starts a global proxy for the entire test suite: 
https://github.com/wagmi-dev/anvil.js/blob/b09bd0ec8013a674dcc5dbe6885e12425bd1f823/examples/example-vitest/tests/globalSetup.ts#L1-L14

Then, you would configure your rpc clients (e.g. with `viem`) to use http / ws rpc urls like this: `http://127.0.0.1:8545/1`. Note the `/1` suffix. In a `vitest` setup, you'd not hard-code that to `/1` but to `/${process.env.VITEST_POOL_ID}`. Hence: `http://127.0.0.1:8545/${process.env.VITEST_POOL_ID`. E.g see here:

https://github.com/wagmi-dev/anvil.js/blob/b09bd0ec8013a674dcc5dbe6885e12425bd1f823/examples/example-vitest/tests/utils.ts#L9-L24

And then, you can optionally reset your `anvil` instance after each test file (or whatever suits your setup):

https://github.com/wagmi-dev/anvil.js/blob/b09bd0ec8013a674dcc5dbe6885e12425bd1f823/examples/example-vitest/tests/setup.ts#L6-L12

You can also print the anvil logs if a test failed:

https://github.com/wagmi-dev/anvil.js/blob/b09bd0ec8013a674dcc5dbe6885e12425bd1f823/examples/example-vitest/tests/setup.ts#L14-L21

**Comment by fubhy** [comment]
We should probably link directly to this example from the readme!

**Comment by fubhy** [comment]
I fixed the wording in the readme file.

**Comment by fubhy** [comment]
Viem already has this setup integrated btw. Logs are printed here for instance: https://github.com/wagmi-dev/viem/actions/runs/4810645056/jobs/8563603002#step:4:926

**Reply by fubhy** [reply]
https://github.com/wagmi-dev/viem/blob/main/src/_test/setup.ts#L40-L59

**Comment by typedarray** [comment]
Thanks, this is very helpful. Just noticed an added benefit of using anvil.js - no longer need to remember to run `anvil` before running `pnpm test` 😁
